source/differentiation/ad/ad_helpers.cc

```diff
@@ -0,0 +1,783 @@
+
+      out = independent_variable_values[index];
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    unsigned int
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::
+      n_touched_independent_variables() const
+    {
+      return std::accumulate(touched_independent_variables.begin(),
+                             touched_independent_variables.end(),
+                             0u);
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    std::size_t
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::n_independent_variables()
+      const
+    {
+      return independent_variable_values.size();
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    unsigned int
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::
+      n_touched_dependent_variables() const
+    {
+      return std::accumulate(touched_dependent_variables.begin(),
+                             touched_dependent_variables.end(),
+                             0u);
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    std::size_t
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::n_dependent_variables()
+      const
+    {
+      return dependent_variables.size();
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    bool
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::is_recording() const
+    {
+      return is_recording_flag;
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    types::tape_index
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::active_tape() const
+    {
+      return active_tape_index;
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    bool
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::is_registered_tape(
+      const types::tape_index &tape_index) const
+    {
+      return registered_tapes.find(tape_index) != registered_tapes.end();
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    void
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::print(
+      std::ostream &stream) const
+    {
+      stream << "Active tape index: " << active_tape_index << "\n";
+      stream << "Registered tapes: ";
+      for (unsigned int i = 0; i < registered_tapes.size(); ++i)
+        stream << *(std::next(registered_tapes.begin(), i))
```

Comment:
That's an `n^2` algorithm here. I suspect that that doesn't matter because we will never have a significant number of tapes registered, right?

Reply:
Thanks for pointing this out. Yes, there should not be many tapes in existence. But I've fixed this nonetheless.